packages/headless/src/controllers/folded-result-list/headless-folded-result-list.ts

```diff
@@ -117,12 +122,32 @@ export function buildFoldedResultList(
   return {
     ...controller,
 
+    loadCollection: (collection) =>
+      dispatch(
+        loadCollection(
+          collection.raw[engine.state.folding.fields.collection] as string
+        )
+      ),
+
     get state() {
       const state = getState();
 
       return {
         ...controller.state,
-        results: state.folding.collections,
+        results: controller.state.results.map((result) => {
+          const collectionId = result.raw[state.folding.fields.collection] as
+            | string
+            | undefined;
+          if (!collectionId || !state.folding.collections[collectionId]) {
+            return {
+              ...result,
+              moreResultsAvailable: false,
+              isLoadingMoreResults: false,
+              children: [],
+            };
+          }
+          return state.folding.collections[collectionId];
+        }),
```

Comment:
Following [this comment](https://github.com/coveo/ui-kit/pull/791/files#r629471003) I decided to change `state.folding.collections` from a `Collection[]` into a `Record<CollectionId, Collection>`.

In doing this, I discovered an issue I hadn't noticed before. It is documented that [not all results need to include folding fields](https://docs.coveo.com/en/1884/searching-with-coveo/understanding-result-folding#:~:text=don%E2%80%99t%20configure%20these%20fields%20for%20item%20types%20you%20don%E2%80%99t%20want%20to%20see%20folded.), which adds more logic here.

An alternative to having all this logic here would be to transform `state.folding.collections` back into an array which would contain all results in the correct order, but would be occasionally redundant with `state.search.results`.

Reply:
Before choosing the technical implementation, I want to make sure we are getting the api right and offering the best developer experience.

If extending an interface is not the best experience because `ResultListState['result']` and `FoldedResultListState['result']` need to have different structures, there are different ways we can work-around this constraint.

If the ResultList controller did not exist, what api design would you expose?

Reply:
I would be trying to expose fetchMoreResults, fieldsToInclude, etc. In JSUI, Folding was a component that modified search results because apart from replacing root results and giving them children it behaved pretty much entirely the same.

Honestly, we could remove folded-result-list altogether and just have folding be an option in result-list.

Reply:
Could there be other draw-backs to switching to an option on the result list?
- The folding fields become required because if we placed defaults, we would always fold?
- The state would expose two properties, `result` and `collection`, even though a developer only needs to use one at a time, which I think is confusing? 
- The `loadCollection` method would only be useful if the folding option is specified.

Correct me if I am wrong, but it feels like combining would be mixing two concerns and make the dev experience more difficult?

Rather than extending the `ResultList` interface, we could copy just the methods that we want, and define `state` with a different type. In the FoldedResultList controller, we would still be free to reuse the actual ResultList controller implementations.

Reply:
> The folding fields become required because if we placed defaults, we would always fold?

We could add a `foldingEnabled` option in the `folding` state.

> The state would expose two properties, `result` and `collection`, even though a developer only needs to use one at a time, which I think is confusing?

I'm not sure what you mean by this. In the current implementation, collections are just results with extra fields, there's no redundancy. I did it this way so I could re-use methods from the result list controller seamlessly.

> The `loadCollection` method would only be useful if the folding option is specified.

I think that would be one of the main drawbacks. `loadCollection` would be useless when folding is disabled. We could make it log an error when folding is disabled, but that wouldn't change that there's a useless method suggested by intellisense.

> Correct me if I am wrong, but it feels like combining would be mixing two concerns and make the dev experience more difficult?

I think where the concerns would be the most mixed, and the main drawback, would be that, if we decided to have a single controller, we would have to make one of the following compromises:
* Remove `FoldedCollection` and add the `children` property to `Result`, making it no longer representative of the API's response.
* Introduce a new `Result` interface to represent the API response and add `children` to the `Result` interface used by the controller.
* Replace the type of `results` in `ResultList` from `Result[]` to `FoldedCollection[]`, ruining the developer experience and breaking existing implementations. In this case, we would also have to rename `FoldedCollection` to something more relevant to results. I would be strongly against this.

Given the necessity of one of those compromises, I changed my mind and no longer think that we should merge folded-result-list and result-list.

> Rather than extending the `ResultList` interface, we could copy just the methods that we want, and define `state` with a different type. In the FoldedResultList controller, we would still be free to reuse the actual ResultList controller implementations.

Thinking back, I'm not sure why attempting to do this didn't seem to work. I could try it again. 

However, I'm still not sure what we gain from doing this. The current implementation that extends `Result` has the advantage that somebody could replace their `ResultList` controller with a `FoldedResultList` controller without having to make any major change to their implementation.

Reply:
> However, I'm still not sure what we gain from doing this.

Isn't it to be able to have `Result` be a [different shape](https://github.com/coveo/ui-kit/pull/824#discussion_r635367758)? Not be restricted to the `Result` interface/extend from it. A developer would see that there is a root result and children instead of:

```
Result {
    ...properties,
    children,
    ...properties
}
```

Reply:
I guess it does reduce confusion a bit in the sense that it divides what is returned by the API vs what was processed, but my reflex is to think those things should be abstracted away. However, given that right now `Result` represents the API, I'll separate them and maybe in a future major version we could refactor `Result`.